File: src/main/java/com/karmanno/plugins/tasks/AssignTagTask.java
package com.karmanno.plugins.tasks;

import com.karmanno.plugins.services.PrepareVersionService;
import com.karmanno.plugins.utils.GitUtils;
import lombok.SneakyThrows;
import org.eclipse.jgit.api.Git;
import org.eclipse.jgit.lib.Constants;
import org.eclipse.jgit.revwalk.RevWalk;
import org.gradle.api.DefaultTask;
import org.gradle.api.Project;
import org.gradle.api.tasks.TaskAction;

public class AssignTagTask extends DefaultTask {

    @TaskAction
    @SneakyThrows
    public void doTask() {
        Project project = getProject();
        Git git = GitUtils.gitRepo(project);
        String version = new PrepareVersionService().getVersion(git).printVersion();
        project.setVersion(version);
        git.tag().setObjectId(
                new RevWalk(git.getRepository()).parseCommit(
                        git.getRepository().resolve(Constants.HEAD)
                )
        ).setName(version).call();
    }
}

File: src/main/kotlin/com/karmanno/plugins/task/AssignTagTask.kt
package com.karmanno.plugins.task

import org.eclipse.jgit.lib.Constants
import org.eclipse.jgit.revwalk.RevWalk
import org.gradle.api.DefaultTask
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.Optional
import org.gradle.api.tasks.TaskAction
import org.gradle.api.tasks.options.Option

open class AssignTagTask: DefaultTask() {

    @Input
    @Optional
    @Option(description = "Release branch, if it is different to master", option = "releaseBranch")
    var releaseBranch: String? = null


    @TaskAction
    fun doTask() {
        val (git, nextVersion) = GenericIncreaseVersionTask().calculateVersionString(project, releaseBranch)

        val head = RevWalk(git.repository).parseCommit(
            git.repository.resolve(Constants.HEAD)
        )

        git.tag()
            .setObjectId(head)
            .setName(nextVersion)
            .setTagger(head.authorIdent)
            .call()
    }
}